### DISCUSSION TITLE
Multiline non-editable text box

### DISCUSSION THREAD

**Discussion by RichoDemus** [emilk/egui#322] [Q&A]
Hi! So if I want to a bigger section of text, like a paragraph, which would be non-editable (selectable is ok), how would I go about that? is that also a Label? Labels kinda make me think of shorter single-line few-word texts.

**Comment by RichoDemus** [comment]
oh, and ideally scrollable!

**Reply by emilk** [reply]
A label can have multiple lines. Just put it in a `ScrollArea` if you want scrolling.

The text in a label isn't selectable though. In most cases I think it is better with a separate "click to copy" button anyway (📋)

**Reply by kotx** [reply]
> The text in a label isn't selectable though. In most cases I think it is better with a separate "click to copy" button anyway (📋)

What if the user wants to copy a specific section of the text? If the paragraph is very large.

**Reply by emilk** [reply]
I have now added support for this: https://github.com/emilk/egui/commit/8d854391df39ebf10ad611edbc432889fc1d1d78

**Reply by wlwatkins** [reply]
is there an example on how it is implemented, i cannot seem to understand the docs (quite new to rust) thanks

**Reply by hacknus** [reply]
Yes, I'm also having trouble to set the mutability. for me `is_mutable` always returns true for the TextBuffer, makes sense, since we are only allowed to pass in mutable strings in `singleline` or `multiline`. I'd be happy to try to help to get fix this in the library, but I'm not entirely sure how this was meant to be implemented?

even if I set the `is_mutable = false` in the `builder.rs` (line 337)
```rust
pub fn show(self, ui: &mut Ui) -> TextEditOutput {
    let is_mutable = false;
    ...
}
```
the text is still mutable by the user - but not focused anymore, so halfway there.

**Reply by NickNick** [reply]
You can pass in a reference to an immutable:

```
let x: &str = "Hello\nWorld!";
ui.text_edit_multiline(&mut x);
```

**Reply by st4s1k** [reply]
> is there an example on how it is implemented, i cannot seem to understand the docs (quite new to rust) thanks

if you have a string, then `&mut my_string.as_str()`, it's a mutable TextBuffer for a `&str` and:
```rs
impl<'a> TextBuffer for &'a str {
    fn is_mutable(&self) -> bool {
        false
    }
    //...
}
```

which is kinda weird, I wish I could pass an argument, or do something like this:
```rs
TextEdit::multiline(&mut my_string)
                    .read_only()
                    .is_editable(false)
                    .is_mutable(false)
```
